pubnub-kotlin/pubnub-kotlin-api/src/commonMain/kotlin/com/pubnub/api/utils/TimetokenUtil.kt

```diff
@@ -0,0 +1,89 @@
+private const val MAXIMUM_TIMETOKEN_VALUE = 99_999_999_999_999_999
+
+private const val NUMBER_OF_DIGITS_IN_TIMETOKEN = 17
+
+/**
+ * Utility object for converting PubNub timetokens to various date-time representations and vice versa.
+ *
+ * This utility provides methods for converting between PubNub timetokens, Unix timestamps, and date-time objects.
+ * A PubNub timetoken is a 17-digit number that represents the number of 100-nanosecond intervals since
+ * January 1, 1970 (UTC). These methods allow for easy conversion between different time representations
+ * used in PubNub, Unix, and standard date-time formats.
+ */
+object TimetokenUtil {
+    /**
+     * Converts a PubNub timetoken (a unique identifier for each message sent and received in a PubNub channel that is
+     * a number of 100-nanosecond intervals since January 1, 1970) to LocalDateTime object representing
+     * the corresponding date and time.
+     *
+     * @param timetoken PubNub timetoken
+     * @return [Instant] representing the corresponding data and time.
+     * @throws IllegalArgumentException if the timetoken does not have 17 digits.
+     */
+    fun timetokenToInstant(timetoken: Long): Instant {
+        if (isLengthDifferentThan17Digits(timetoken)) {
+            throw IllegalArgumentException("Timetoken should have 17 digits")
+        }
+        // Convert timetoken to seconds and nanoseconds components
+        val epochSeconds = timetoken / 10_000_000 // Divide by 10^7 to get seconds
+        val epochNanoseconds =
+            ((timetoken % 10_000_000) * 100).toInt() // The remainder, multiplied by 100 to get nanoseconds
+
+        return Instant.fromEpochSeconds(epochSeconds, epochNanoseconds)
+    }
+
+    /**
+     * Converts [Instant] to a PubNub timetoken
+     *
+     * A PubNub timetoken is a 17-digit number representing the number of 100-nanosecond intervals since January 1, 1970.
+     *
+     * @param instant The [Instant] object to be converted to a PubNub timetoken.
+     * @return A 17-digit [Long] representing the PubNub timetoken for the given [Instant].
+     */
+    fun instantToTimetoken(instant: Instant): Long {
+        val epochSeconds: Long = instant.epochSeconds
+        val epochNanoseconds: Int = instant.nanosecondsOfSecond
+
+        val pnTimetokenInNanosecondsWithoutNanoPrecision = epochSeconds * 10_000_000
+        val nanosecondsForPubNubTimetoken =
+            epochNanoseconds / 100 // PubNub timetoken for nanoseconds store only 7 digits instead of 9
+        return pnTimetokenInNanosecondsWithoutNanoPrecision + nanosecondsForPubNubTimetoken
+    }
+
+    /**
+     * Converts a Unix timestamp to a PubNub timetoken
+     *
+     * @param unixTime The Unix timestamp to be converted to a PubNub timetoken.
+     * @return A 17-digit [Long] representing the PubNub timetoken corresponding to the given Unix timestamp.
+     */
+    fun unixToTimetoken(unixTime: Long): Long {
+        val numberOfDigits = unixTime.toString().length
```

Comment:
why not just `= unixTime * 10_000` ?

Reply:
as of now this test won't pass:

```
    @Test
    fun conversion_is_reversible() {
        val unixTimes = listOf(10, 100, 1000, 10_000, 100_000, 1_000_000_000_000)
        unixTimes.forEach {
            assertEquals(it, TimetokenUtil.timetokenToUnix(TimetokenUtil.unixToTimetoken(it)))
        }
    }

```

Reply:
I added validation and description that unix timetoken should has 10 or 13 digits.

Conversion:
assertEquals(it, TimetokenUtil.timetokenToUnix(TimetokenUtil.unixToTimetoken(it)))
is reversible only when unix timetoken has 13 digits because TimetokenUtil.timetokenToUnix returns 13 digits.

What do you think?

Reply:
then if we want to have this limitation that we need 13 digits, I think the `unixToTimetoken` function should **only** accept milliseconds.

As it is, the function which could be just `* 10000` is way too complicated... it's using string conversions, counting characters, doubles, pow() and so on... and all that to handle milliseconds or seconds, which IMHO makes it just more confusing. 

I would keep it simple both from the perspective of what the function does (i.e. converts milliseconds to timetokens) as well as the implementation side. If you think we should have a function that takes seconds, it should be a separate function, but IMHO it's not needed

Reply:
What do you think about:
```
    fun unixToTimetoken(unixTime: Long): Long {
        val unixTimeLength = unixTime.toString().length

        return when (unixTimeLength) {
            UNIX_TIME_IN_SECONDS -> { unixTime * 1_000_000 }
            UNIX_TIME_IN_MILLISECONDS -> { unixTime * 1_000 }
            else -> { throw IllegalArgumentException("Unix timetoken should have 10 or 13 digits") }
        }
    }
```

Reply:
this only resolves the concerns about the implementation (though the string conversion is still unnecessary, I'd recommend just comparing MIN > unixTime > MAX to see if it's in the valid range)

but honestly I don't think this function can take two different units and magically return the correct result, because there's no way to fix the reverse function - `timetokenToUnix` cannot guess which unit you want to get back. 

The only way this could work with seconds AND milliseconds (and any other unit) is by adding a `unit: DurationUnit` parameter to both `unixToTimetoken` and `timetokenToUnix` and then the user can pass whichever unit they expect.

Otherwise this is just too confusing and error-prone

Reply:
for example: I have a class `MyTimestampUtility` with methods like `saveTimestamp()` and `getTimestamp()` . It operates on *seconds* (and not *milliseconds*).

I use `getTimestamp()` in seconds, then pass it to `getHistory(from = unixToTimetoken(seconds))`, and I get some messages back. Now I select the last message I'm interested in and want to save its timestamp back to my object.
So naturally I do `myTimestampUtility.saveTimestamp( timetokenToUnix(message.timetoken) )`... which is wrong

can you see  why I really don't like it?

Reply:
I think I see your point. For me the fact that `timetokenToUnix` returns only milliseconds is fine, this fact is documented. Removing seconds handling from `unixToTimetoken` cause that users operating on seconds have more work todo.

I will remove seconds handling from `unixToTimetoken` method.